test/spec/dom-utils.spec.js

```diff
@@ -269,24 +269,97 @@ describe( 'DOM utils', () => {
             expect( getXPath( node, null ) ).to.equal( '/root/path/to/node' );
         } );
 
+        it( 'returns same /root/path/to/node if first parameter is null and second parameter is true', () => {
+            const node = xml.querySelector( 'node' );
+            expect( getXPath( node, null ) ).to.equal( '/root/path/to/node' );
+        } );
+
         it( 'returns path from context first node provided as parameter', () => {
             const node = xml.querySelector( 'node' );
             expect( getXPath( node, 'root' ) ).to.equal( '/path/to/node' );
         } );
+
         it( 'returned path includes no positions if there are no siblings with the same name along the path', () => {
             const node = xml.querySelector( 'node' );
             expect( getXPath( node, 'root', true ) ).to.equal( '/path/to/node' );
         } );
+
         it( 'returned path includes positions when asked', () => {
             const node = xml.querySelectorAll( 'number' )[ 1 ];
             expect( getXPath( node, 'root', true ) ).to.equal( '/path/to/repeat[2]/number' );
         } );
+
+        it( 'returned path includes position of first repeat if sibling repeat exists', () => {
+            const node = xml.querySelector( 'number' );
+            expect( getXPath( node, 'root', true ) ).to.equal( '/path/to/repeat[1]/number' );
```

Comment:
Good points. Thanks! I'd vote for keeping `node` as the first argument, to cleanly be able to call `getXPath(node)`.   

I can see the second argument is confusing. It's not quite _relative to nodeName_. The reason it exists is that it was/is important to be able to pass the whole `<model>` to the native XPath evaluator (root=model) but still consider the first child of the first instance as the root for XPath expressions. So we usually use `getXPath(node, 'instance') to get `/data/node` paths. "rootNodeName" is an incorrect name though which I hadn't realized until now - should perhaps be something like stopBeforeNode or something like that.

The reason we want(ed) to pass the whole `<model>` to the native XPath evaluator is that this enables us to also evaluate large secondary instances natively by prepending all paths (shiftRoot in form-model.js). This is one of the slowest actions.

This has become less important with new hybrid evaluator though (which could implement the `instance(id)` function, though we'd have to add some new parameter to pass secondary instances). I think it would still be worthwhile to send certain expressions to the native evaluator but am not sure if this was actually measured

Reply:
> I'd vote for keeping node as the first argument, to cleanly be able to call getXPath(node).

Agree. My thinking with the separate functions was that `getXPath` is operating on `node` as data context-free (or as you say, in the scope of the whole `<model>`). Whereas the hypothetical `getRelativeXPath` is scoped to `rootNodeName` (actually `rootNode` with its name as a selector shorthand to reference it without selecting it) with `node` as the relevant data it needs to perform the operation. With the mindset that generally context-first/data-last tends to be more composable/reusable.

That said, I think the separate functions approach is probably less desirable now that I've thought on it for a while. The rest of this context is helpful for thinking about this in the future 🙂  

> This has become less important with new hybrid evaluator though (which could implement the instance(id) function, though we'd have to add some new parameter to pass secondary instances). I think it would still be worthwhile to send certain expressions to the native evaluator but am not sure if this was actually measured

Not blocking my review (none of this thread is), but I'm curious to know more about this, specifically about what considerations go into using which evaluator.